src/ruby/puppetserver-lib/puppet/server/ast_compiler.rb

```diff
@@ -0,0 +1,44 @@
+require 'puppet_pal'
+require 'puppet/server/log_collector'
+
+module Puppet
+  module Server
+    class ASTCompiler
+      def self.compile(compile_options)
+        options = compile_options['options'] || {}
+
+        log_level = options['log_level']
+
+        if options['capture_logs']
+          catalog, logs = Logging.capture_logs(log_level) do
+            compile_ast(compile_options)
+          end
+
+          { catalog: catalog, logs: logs }
+        else
+          catalog = compile_ast(compile_options)
+          { catalog: catalog }
+        end
+      end
+
+      def self.compile_ast(compile_options)
+        code = compile_options['code_ast']
+        target = compile_options['certname']
+
+        Puppet::Pal.in_tmp_environment(compile_options['environment'],
```

Comment:
Do we have to use Pal?

Reply:
I don't know that we _have_ to. I was going to try to get it working using PAL first, since we have an example of that flow, and when we looked over this code, I don't think anyone had a good reason _not_ to. Its API is definitely better defined for what we're trying to do than the other entry points.

Reply:
That being said, now that this is working, I'll start looking at alternatives, if we think this is gross or unnecessary. Though I think we might wind up with more complicated code here.

Do you have some particular things in mind you'd like to avoid? I've cut out almost all of the settings configuration already, compared to Bolt.

Reply:
I just don't know what this does or is meant to do.